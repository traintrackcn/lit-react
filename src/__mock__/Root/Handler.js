import LITComponentHandler from '../../LITComponentHandler';
import StateHandler from './StateHandler';
import p from '../p';

export default class RootHandler extends LITComponentHandler {


    getStateListenerTarget() {
        return this.getProperty1();
    }

    setProperty1(value) {
        const h = this.getStateHandler();
        return h.setProperty1(value);
    }

    getProperty1() {
        const h = this.getStateHandler();
        return h.getProperty1();
    }

    getStateHandler() {
        if (!this._state) {
            this._state = new StateHandler(p, this);
        }
        return this._state;
    }

    getHandler() {
        return this;
    }

}