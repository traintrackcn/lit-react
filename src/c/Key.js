export default class {

    constructor(pValue, value){
        
        const list = [];
        if (pValue) list.push(pValue);
        if (value) list.push(value);

        if(list.length) this.value = list.join(',');
        
    }

    isEqual(anotherK){
        if (this.value && anotherK){
            if (this.value === anotherK.value) return true;
        }
        return false;
    }

}