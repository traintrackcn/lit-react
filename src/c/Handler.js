import utils from './_utils';
import Node from './Node';
import Key from './Key';

export default class {

    constructor(){
        this.map = {};
        this.cache = {};
    }

    parse(data){
        this.map = utils.getTreeMap(data);
    }

    get(key, ctx){
        const node = this.get_node(key);
        return node.get(ctx);
    }

    get_k(key){
        var k;
        if (key && ( key instanceof Key)){
            k = key.value;
        }else{
            k = key;
        }
        return k;
    }   

    get_raw_node(key){

        const k = this.get_k(key);
        return this.map[k];
    }


    get_node(key){
        const k = this.get_k(key);
        var cached = this.cache[k];
        if (cached) return cached;

        if (!this.map) return;

        const node = this.map[k];
        console.log('key ->', k, node);
        cached = new Node(node);

        //set to cache
        this.cache[k] = cached;

        return cached;
    }


}