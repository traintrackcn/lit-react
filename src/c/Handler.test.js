import Handler from "./Handler";
import Key from "./Key";

describe('', () => {
    const handler = new Handler();
    beforeAll(()=>{
        handler.parse(raw);
    });

    it('current map', () => {
        expect(handler.map).toMatchSnapshot();
    });

    it('get raw node "config,menu"', () => {
        const key = new Key();
        key.value = 'config,menu';
        expect(handler.get_raw_node(key)).toMatchSnapshot();
        expect(handler.get_raw_node('config,menu')).toMatchSnapshot();
    });

    it('get node "config,menu"', () => {
        const key = new Key();
        key.value = 'config,menu';
        expect(handler.get_node(key)).toMatchSnapshot();
        expect(handler.get_node('config,menu')).toMatchSnapshot();
    });

    it('get node "config,menu,undefined"', () => {
      const key = new Key();
      const kValue = 'config,menu,undefined';
      key.value = kValue;
      expect(handler.get_node(key)).toMatchSnapshot();
      expect(handler.get_node(kValue)).toMatchSnapshot();
  });

    it('get node "config,menu,web" base rule', () => {
        const key = new Key();
        key.value = 'config,menu,WEB';
        expect(handler.get(key)).toMatchSnapshot();
        expect(handler.get('config,menu,WEB')).toMatchSnapshot();


        const ctx = ['prj1'];
        expect(handler.get(key, ctx)).toMatchSnapshot();
    });

});



const raw = [
    {
      "title": "menu",
      "expanded": true,
      "children": [
        {
          "title": "SERVICE",
          "active": true
        },
        {
          "title": "BOT",
          "active": true
        },
        {
          "title": "CONFIG",
          "expanded": true,
          "active": true
        }
      ]
    },
    {
      "title": "config",
      "expanded": true,
      "children": [
        {
          "title": "menu",
          "expanded": false,
          "children": [
            {
              "title": "WEB",
              "r": [
                {
                  "on": true,
                  "base": "LIT"
                },
                {
                  "_": [
                    "prj1"
                  ],
                  "base": "OTHER"
                }
              ],
              "active": true
            },
            {
              "title": "APP",
              "r": [
                {
                  "on": true,
                  "base": "LIT"
                }
              ],
              "expanded": true,
              "active": true
            },
            {
              "title": "API",
              "active": true
            },
            {
              "title": "MISC",
              "active": true
            }
          ]
        }
      ]
    }
];