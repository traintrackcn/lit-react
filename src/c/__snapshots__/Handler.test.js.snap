// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current map 1`] = `
Object {
  "config": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "active": true,
            "r": Array [
              Object {
                "base": "LIT",
                "on": true,
              },
              Object {
                "_": Array [
                  "prj1",
                ],
                "base": "OTHER",
              },
            ],
            "title": "WEB",
          },
          Object {
            "active": true,
            "expanded": true,
            "r": Array [
              Object {
                "base": "LIT",
                "on": true,
              },
            ],
            "title": "APP",
          },
          Object {
            "active": true,
            "title": "API",
          },
          Object {
            "active": true,
            "title": "MISC",
          },
        ],
        "expanded": false,
        "title": "menu",
      },
    ],
    "expanded": true,
    "title": "config",
  },
  "config,menu": Object {
    "children": Array [
      Object {
        "active": true,
        "r": Array [
          Object {
            "base": "LIT",
            "on": true,
          },
          Object {
            "_": Array [
              "prj1",
            ],
            "base": "OTHER",
          },
        ],
        "title": "WEB",
      },
      Object {
        "active": true,
        "expanded": true,
        "r": Array [
          Object {
            "base": "LIT",
            "on": true,
          },
        ],
        "title": "APP",
      },
      Object {
        "active": true,
        "title": "API",
      },
      Object {
        "active": true,
        "title": "MISC",
      },
    ],
    "expanded": false,
    "title": "menu",
  },
  "config,menu,API": Object {
    "active": true,
    "title": "API",
  },
  "config,menu,APP": Object {
    "active": true,
    "expanded": true,
    "r": Array [
      Object {
        "base": "LIT",
        "on": true,
      },
    ],
    "title": "APP",
  },
  "config,menu,MISC": Object {
    "active": true,
    "title": "MISC",
  },
  "config,menu,WEB": Object {
    "active": true,
    "r": Array [
      Object {
        "base": "LIT",
        "on": true,
      },
      Object {
        "_": Array [
          "prj1",
        ],
        "base": "OTHER",
      },
    ],
    "title": "WEB",
  },
  "menu": Object {
    "children": Array [
      Object {
        "active": true,
        "title": "SERVICE",
      },
      Object {
        "active": true,
        "title": "BOT",
      },
      Object {
        "active": true,
        "expanded": true,
        "title": "CONFIG",
      },
    ],
    "expanded": true,
    "title": "menu",
  },
  "menu,BOT": Object {
    "active": true,
    "title": "BOT",
  },
  "menu,CONFIG": Object {
    "active": true,
    "expanded": true,
    "title": "CONFIG",
  },
  "menu,SERVICE": Object {
    "active": true,
    "title": "SERVICE",
  },
}
`;

exports[`get node "config,menu" 1`] = `
_default {
  "active": false,
  "base": undefined,
  "children": Array [
    Object {
      "active": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
        Object {
          "_": Array [
            "prj1",
          ],
          "base": "OTHER",
        },
      ],
      "title": "WEB",
    },
    Object {
      "active": true,
      "expanded": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
      ],
      "title": "APP",
    },
    Object {
      "active": true,
      "title": "API",
    },
    Object {
      "active": true,
      "title": "MISC",
    },
  ],
  "description": undefined,
  "rules": undefined,
  "title": "menu",
}
`;

exports[`get node "config,menu" 2`] = `
_default {
  "active": false,
  "base": undefined,
  "children": Array [
    Object {
      "active": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
        Object {
          "_": Array [
            "prj1",
          ],
          "base": "OTHER",
        },
      ],
      "title": "WEB",
    },
    Object {
      "active": true,
      "expanded": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
      ],
      "title": "APP",
    },
    Object {
      "active": true,
      "title": "API",
    },
    Object {
      "active": true,
      "title": "MISC",
    },
  ],
  "description": undefined,
  "rules": undefined,
  "title": "menu",
}
`;

exports[`get node "config,menu,web" base rule 1`] = `
Object {
  "base": "LIT",
  "on": true,
}
`;

exports[`get node "config,menu,web" base rule 2`] = `
Object {
  "base": "LIT",
  "on": true,
}
`;

exports[`get node "config,menu,web" base rule 3`] = `
Object {
  "base": "OTHER",
  "ctx": Array [
    "prj1",
  ],
  "on": true,
}
`;

exports[`get raw node "config,menu" 1`] = `
Object {
  "children": Array [
    Object {
      "active": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
        Object {
          "_": Array [
            "prj1",
          ],
          "base": "OTHER",
        },
      ],
      "title": "WEB",
    },
    Object {
      "active": true,
      "expanded": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
      ],
      "title": "APP",
    },
    Object {
      "active": true,
      "title": "API",
    },
    Object {
      "active": true,
      "title": "MISC",
    },
  ],
  "expanded": false,
  "title": "menu",
}
`;

exports[`get raw node "config,menu" 2`] = `
Object {
  "children": Array [
    Object {
      "active": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
        Object {
          "_": Array [
            "prj1",
          ],
          "base": "OTHER",
        },
      ],
      "title": "WEB",
    },
    Object {
      "active": true,
      "expanded": true,
      "r": Array [
        Object {
          "base": "LIT",
          "on": true,
        },
      ],
      "title": "APP",
    },
    Object {
      "active": true,
      "title": "API",
    },
    Object {
      "active": true,
      "title": "MISC",
    },
  ],
  "expanded": false,
  "title": "menu",
}
`;
